saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/authentication/Saml2AuthenticationRequest.java

```diff
@@ -167,6 +208,39 @@ public Builder assertionConsumerServiceUrl(String assertionConsumerServiceUrl) {
 			return this;
 		}
 
+		/**
+		 * Sets the desired binding and encoding for the authentication request.
+		 * If {@code binding==REDIRECT} then DEFLATE encoding will be used.
+		 * The authentication request will contain {@code SigAlg} and {@code Signature} parameters.
+		 * If {@code binding!=REDIRECT} then XML Signatures will be used if they are signing credentials available.
+		 * The authentication request will contain {@code SigAlg} and {@code Signature} parameters.
+		 * @param binding - the binding that the message should be encoded and sent for.
+		 * @return this {@code Builder}
+		 */
+		public Builder binding(Saml2MessageBinding binding) {
+			this.binding = binding;
+			return this;
+		}
+
+		/**
+		 * Modifies the collection of unencoded {@code Query Parameters} used in the AuthNRequest
+		 * For example:
+		 * <code>
+		 *     String relayState = ...;
+		 *     return Saml2AuthenticationRequest.withLocalSpEntityId("id")
+		 *             .queryParameters(qp -> qp.put("RelayState", relayState))
+		 *             ...
+		 *             .build();
+		 * </code>
+		 * All parameters should NOT be encoded in the map, as they may become double encoded.
+		 * @param queryParameters - a consumer that can modify the collection of unencoded query parameters
+		 * @return this object
+		 */
+		public Builder queryParameters(Consumer<MultiValueMap<String, String>> queryParameters) {
```

Comment:
I think this would be better having individual methods for spec-related parameters like `RelayState` and `SAMLRequest`.

I think that this method still has its place, `relayState(String relayState)`, for example, put simply call `this.queryParameters.put("RelayState", relayState)`

Reply:
I agree.
Addressed in: cd4c9d1198e1cd7c800fc92e10c488902f882d6a